### ISSUE TITLE
Remove time and mem usage calls

### ISSUE THREAD

**Issue by linas** [opencog/link-grammar#1020]
Following discussion on #1019:

on my system, some of my tests spend 5% to 15% cpu in system time, instead of user. This makes no sense to me, as LG makes almost no system calls ... except for some "infrequent" calls to get the clock (for timeouts). Perhaps we are checking the timer much too often? Hard to say.

Propose -- Use a timer interrupt. That way, we can eliminate all calls to check the elapsed time. .. and if we're doing that, we may as well remove the calls to check mem usage.

Some notes about RAM usage:
1) RAM is no longer a constraint on modern systems (the original LG code dates to 1995, when it was.)
2) In the highly unlikely event that LG runs on a memory-constrained system (100's of MB or less) then having a special malloc library for dealing with mem constraint makes more sense.  Also, rather than checking for `nullptr == malloc()` I instead I envision such a special lib would throw an exception which LG would need to catch and cleanup. Which  would require new code & design, but really I strongly doubt that anyone will ever run LG on a small RAM system.  

So basically, this proposal is to remove `resources.c` and `resources.h` entirely.

**Comment by ampli** [comment]
> Perhaps we are checking the timer much too often? Hard to say.

It turns out this is an old problem in `post_process_lkgs()` and `prepare_to_parse()`.

In `post_process_lkgs()`:
``` c
1176	if ((63 == in%64) && resources_exhausted(opts->resources)) break;
...
1203  if ((15 == in%16) && resources_exhausted(opts->resources)) break;
```
**For now, these numbers should just be increased.**
Eventually these checks there should be dropped altogether - when I finish my total rewrite of post processing. The current code there is written in a very wasteful way and I hope to get there 3 order of magnitude less CPU, and maybe move part of it to the pruning stage (I think I found how to implement that), a thing that will solve other problems too.

In  `prepare_to_parse()`:
``` c
    150         for (i=0; i<sent->length; i++)
    151         {
    152                 sent->word[i].d = eliminate_duplicate_disjuncts(sent->word[i].d);
    153 
    154                 /* Some long Russian sentences can really blow up, here. */
    155                 if (resources_exhausted(opts->resources))
    156                         return;
    157         }
```
So `resources_exhausted()` is needlessly called per token. Since the library can process 1000's tokens per second the timer call rate there is high. It is now (due to a new implementation) not possible for `eliminate_duplicate_disjuncts()` to take much time (unless maybe if I increase MAX_SENTENCE to 4095..., in which case a solution like `checktimer` can be used).

**I will change it to one call after the loop.**

In `count.c` there is absolutely no problem (it turns out the slow panic resolution is not related):
Try to uncomment the following `fprintf()` in `count.c`:
``` c
    186         if (ctxt->exhausted || ((0 == ctxt->checktimer%(1<<22)) &&
    187                                (ctxt->current_resources != NULL) &&
    188                                //fprintf(stderr, "T") &&
    189                                resources_exhausted(ctxt->current_resources)))
```
You will see that it prints `T` in a very low rate. So this actually gets handled well by the high value of `checktimer`'s divisor.

Additional timer calls are done by print_times() when it doesn't print (e.g. verbosity=1), and this adds several timer calls per sentence. **This can also be fixed.**

> Propose -- Use a timer interrupt. That way, we can eliminate all calls to check the elapsed time. .. 

You really need per thread CPU (not elapsed) timer interrupts. But the real problem is a portability one.
On Linux this is available using `timer_create()`. On other OS types there may not be much luck.
For example on Solaris this needs a special privilege which is usually disabled. Nothing similar to `timer_create()` on Windows (and there is more complexity in that `link-parser` is a console application).  On iOS (see the following URL)  "timer_create doesn't exist and setitimer doesn't fire on iOS, so we're using a pthreads-based implementation"). A SIGALRM-based solution cannot be used because it interrupts system calls. For all of that see: https://github.com/ghc/ghc/blob/master/rts/posix/Itimer.c .

Fortunately, solving this mess (which is not easy) is not currently needed. What I can do is just to modify the current implementation so it will use negligible system resources, and this by definition will solve the problem. At the same time I will start to gather info on how to portably implement per-thread CPU timer interrupts in case this will be needed in the future.

> and if we're doing that, we may as well remove the calls to check mem usage.

- Currently the RAM usage checks are essentially neutralized. So these checks do nothing and are  just a pure overhead (but it is small). I will fix that.
- I found how to implement `malloc()` failure cleanup and it is in my TODO list.

To sum up:
**I will soon send a PR that solves the current timer overhead.**

**Comment by linas** [comment]
Huh. OK. That sounds good.
